### PR TITLE
KEYCLOAK-16703 The username returned by token introspect endpoint is …

### PR DESCRIPTION
**Pull request by ch219318**
The username returned by token introspect endpoint is null when remove or modify username mapper

### CODE REVIEW

**Comment by ch219318** [comment]
@pedroigor @mposolda 
please review

File: services/src/main/java/org/keycloak/protocol/oidc/AccessTokenIntrospectionProvider.java [hunk 1]
```diff
@@ -56,7 +57,11 @@ public Response introspect(String token) {
             if (accessToken != null) {
                 tokenMetadata = JsonSerialization.createObjectNode(accessToken);
                 tokenMetadata.put("client_id", accessToken.getIssuedFor());
-                tokenMetadata.put("username", accessToken.getPreferredUsername());
+
+                UserModel userModel = session.users().getUserById(realm, accessToken.getSubject());
+                if (userModel != null) {
+                    tokenMetadata.put("username", userModel.getUsername());
```
Comment (mduchrow) [inline services/src/main/java/org/keycloak/protocol/oidc/AccessTokenIntrospectionProvider.java:63]: IMO this is not the right solution. An introspection should reflect what's in the incoming access token. The "username" in the access token could have been set deliberately to a different value than the user's username (e.g. an external unique user name). And exactly that value should be provided in the introspection result.

```suggestion                   
                    if (!tokenMetadata.has("username") && tokenMetadata.has("preferred_username")) {
                      tokenMetadata.put("username", tokenMetadata.get(accessToken.getPreferredUsername()));
                    }                                      
```
With that approach we get the correct value from the access token and as a fallback have compatibility for those who are relying on the current behavior.
BTW: "client_id" should be treated the same way.

Comment (ch219318) [reply services/src/main/java/org/keycloak/protocol/oidc/AccessTokenIntrospectionProvider.java:63]: "tokenMetadata.put("username", tokenMetadata.get(accessToken.getPreferredUsername()))" do you mean "tokenMetadata.put("username", tokenMetadata.get("preferred_username"))" or "tokenMetadata.put("username", accessToken.getPreferredUsername())" ?

Comment (mduchrow) [reply services/src/main/java/org/keycloak/protocol/oidc/AccessTokenIntrospectionProvider.java:63]: Well, hopefully both are returning the same value :-)
tokenMetadata.get() returns a JsonNode, so I thought taking it from the accessToken.getPreferredUsername() is easier.

BTW: If "preferred_username" is also not set, your approach to assign it from UserModel.getUsername() might be OK, as a final fallback.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/TokenIntrospectionTest.java [hunk 2]
```diff
@@ -31,17 +31,12 @@
...
 import org.keycloak.jose.jws.JWSInput;
 import org.keycloak.protocol.oidc.OIDCLoginProtocol;
-import org.keycloak.representations.RefreshToken;
-import org.keycloak.representations.idm.ClientRepresentation;
-import org.keycloak.representations.idm.CredentialRepresentation;
-import org.keycloak.representations.idm.EventRepresentation;
-import org.keycloak.representations.idm.OAuth2ErrorRepresentation;
-import org.keycloak.representations.idm.RealmRepresentation;
-import org.keycloak.representations.idm.UserRepresentation;
+import org.keycloak.representations.idm.*;
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/TokenIntrospectionTest.java:39]: Can you please remove star imports?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/TokenIntrospectionTest.java [hunk 3]
```diff
@@ -59,9 +54,7 @@
 import javax.ws.rs.core.UriBuilder;
 
 import java.io.UnsupportedEncodingException;
-import java.util.ArrayList;
-import java.util.LinkedList;
-import java.util.List;
+import java.util.*;
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/TokenIntrospectionTest.java:57]: Can you please remove star imports?

**Review by mposolda — CHANGES_REQUESTED** [review]
@ch219318 Thanks for the PR. To me, it looks good once star imports are removed. Thanks!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/TokenIntrospectionTest.java (hunk 2 above)
Comment (ch219318) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/TokenIntrospectionTest.java:39]: ok

**Review by mposolda — APPROVED** [review]
@ch219318 Thanks, I plan to merge this once tests are passing.

**Pull request merged**